slsDetectorSoftware/src/Module.cpp

```diff
@@ -2386,93 +2386,158 @@ uint64_t Module::getReceiverCurrentFrameIndex() const {
     LOG(logDEBUG1) << "Sending: ["
                    << getFunctionNameFromEnum(static_cast<detFuncs>(fnum))
                    << "]";
     sendToDetector(fnum, nullptr, 0, nullptr, 0);
 }
 
-template <typename Ret> Ret Module::sendToDetector(int fnum) {
+void Module::sendToDetector(int fnum) {
+    static_cast<const Module &>(*this).sendToDetector(fnum);
+}
+
+template <typename Ret> Ret Module::sendToDetector(int fnum) const {
     LOG(logDEBUG1) << "Sending: ["
                    << getFunctionNameFromEnum(static_cast<detFuncs>(fnum))
                    << ", nullptr, 0, " << typeid(Ret).name() << ", "
                    << sizeof(Ret) << "]";
+    STATIC_ASSERT_ARG(Ret, "sendToDetector")
     Ret retval{};
     sendToDetector(fnum, nullptr, 0, &retval, sizeof(retval));
     LOG(logDEBUG1) << "Got back: " << ToString(retval);
     return retval;
 }
 
+template <typename Ret> Ret Module::sendToDetector(int fnum) {
+    return static_cast<const Module &>(*this).sendToDetector<Ret>(fnum);
+}
+
 template <typename Ret, typename Arg>
-Ret Module::sendToDetector(int fnum, const Arg &args) {
+Ret Module::sendToDetector(int fnum, const Arg &args) const {
     LOG(logDEBUG1) << "Sending: ["
                    << getFunctionNameFromEnum(static_cast<detFuncs>(fnum))
                    << ", " << args << ", " << sizeof(args) << ", "
                    << typeid(Ret).name() << ", " << sizeof(Ret) << "]";
+    STATIC_ASSERT_ARG(Arg, "sendToDetector")
+    STATIC_ASSERT_ARG(Ret, "sendToDetector")
     Ret retval{};
     sendToDetector(fnum, &args, sizeof(args), &retval, sizeof(retval));
     LOG(logDEBUG1) << "Got back: " << ToString(retval);
     return retval;
 }
 
-void Module::sendToDetectorStop(int fnum, const void *args, size_t args_size,
-                                void *retval, size_t retval_size) {
-    static_cast<const Module &>(*this).sendToDetectorStop(fnum, args, args_size,
-                                                          retval, retval_size);
+template <typename Ret, typename Arg>
+Ret Module::sendToDetector(int fnum, const Arg &args) {
+    return static_cast<const Module &>(*this).sendToDetector<Ret>(fnum, args);
 }
 
+//---------------------------------------------------------- sendToDetectorStop
+
 void Module::sendToDetectorStop(int fnum, const void *args, size_t args_size,
                                 void *retval, size_t retval_size) const {
+    // This is the only function that actually sends data to the detector stop
+    // the other versions use templates to deduce sizes and create
+    // the return type
+    preSendArgsCheck(args, args_size, retval, retval_size);
     auto stop = DetectorSocket(shm()->hostname, shm()->stopPort);
     stop.sendCommandThenRead(fnum, args, args_size, retval, retval_size);
     stop.close();
 }
 
-template <typename Arg, typename Ret>
-void Module::sendToDetectorStop(int fnum, const Arg &args, Ret &retval) {
-    LOG(logDEBUG1) << "Sending to Stop: ["
```

Comment:
Why are you removing these prints? Are they moved elsewhere? They are removed in a few other places as well..

Reply:
The prints are in the const version of the function, and the non const version just calls the const version. So all prints should still be there, just not duplicated for const and non const. 

Idea is taken from Scott Meyers Effective C++... This way of casting can be used when non const version is not really modifying anything, just that it's called from a non const context.

Reply:
soundd good :)